nixos/modules/services/web-apps/nextcloud.nix

```diff
@@ -374,6 +381,222 @@ in {
+           <para>This option exists since Nextcloud 21! If older versions are used,
+            this will throw an eval-error!</para>
+          </warning>
+
+          Default locale to use when Nextcloud fails to detect the user's
+          locale.
+
+          Further details about this setting and supported locale codes can be found in the
+          <link xlink:href="https://docs.nextcloud.com/server/stable/admin_manual/configuration_server/language_configuration.html">upstream documentation</link>.
+        '';
+      };
+
+      twofactorEnforced = let
+        twoFactorAuthUpstreamManualUrl = "https://docs.nextcloud.com/server/latest/admin_manual/configuration_user/two_factor-auth.html";
+      in {
+        enabled = mkOption {
+          default = null;
+          type = types.nullOr types.bool;
+          example = true;
+          description = ''
+            ${uiClashWarning}
+
+            Whether to enforce two-factor authentication.
+
+            Further details about this setting and supported locale codes can be found in the
+            <link xlink:href="${twoFactorAuthUpstreamManualUrl}">upstream documentation</link>.
+          '';
+        };
+        groups = mkOption {
+          default = null;
+          type = types.nullOr (types.listOf types.str);
+          example = [ "admin" ];
+          description = ''
+            ${uiClashWarning}
+
+            Enforce two-factor authentication for only users in these named
+            groups.  The groups must already exist in the Nextcloud database.
+
+            Further details about this setting and supported locale codes can be found in the
+            <link xlink:href="${twoFactorAuthUpstreamManualUrl}">upstream documentation</link>.
+          '';
+        };
+        excludedGroups = mkOption {
+          default = null;
+          type = types.nullOr (types.listOf types.str);
+          example = [ "guests" ];
+          description = ''
+            ${uiClashWarning}
+
+            Do <emphasis>not</emphasis> enforce two-factor authentication for
+            users in these named groups.  The groups must already exist in the
+            Nextcloud database.
+
+            Further details about this setting and supported locale codes can be found in the
+            <link xlink:href="${twoFactorAuthUpstreamManualUrl}">upstream documentation</link>.
+          '';
+        };
+      };
+
+      mail = let
+        mailUpstreamManualUrl = "https://docs.nextcloud.com/server/stable/admin_manual/configuration_server/email_configuration.html";
+      in {
+        mode = mkOption {
+          default = null;
+          type = types.nullOr (types.enum [ "smtp" "sendmail" "qmail" ]);
+          example = "smtp";
+          description = ''
+            ${uiClashWarning}
+
+            The mail server mode.
+
+            Further details about this setting can be found in the
+            <link xlink:href="${mailUpstreamManualUrl}">upstream documentation</link>.
```

Comment:
Perhaps define this entire paragraph as a variable rather than just the URL, since it is all repeated.

To make the code even dryer, you could define a function which takes the main paragraph as an argument:

```nix
mainDescription: ''
  ${uiClashWarning}

  ${mainDescription}

  Further details about this setting can be found in the
  <link xlink:href="https://docs.nextcloud.com/server/stable/admin_manual/configuration_server/email_configuration.html">upstream documentation</link>.
''
```

Reply:
I actually wanted to refactor this "Further details about this setting" paragraph from every option description that has it, but I thought it would be out of scope for this PR.  Are you OK with leaving this as is in this changeset if I followup with another PR?

Reply:
Yeah. It's probably better to do the refactoring separately, that way we can get this merged sooner.